### ISSUE TITLE
How to add a new loss function to Detectron2

### ISSUE THREAD

**Issue by ferna11i** [facebookresearch/detectron2#974]
## How to add a new loss function (specifically DIoU) to Detectron2

I have tried L1 and L2 loss for Faster RCNN and have achieved good results. However, the loss function proposed and implemented by the authors of this paper: https://arxiv.org/abs/1911.08287 seem to perform quite well and hence I would like to replicate the loss function in Detectron 2.

The authors have used this Pytorch implementation of Detectron : https://github.com/roytseng-tw/Detectron.pytorch to develop and test the loss functions. 
Their implementation is available at: https://github.com/Zzh-tju/DIoU-pytorch-detectron
The code for each loss function is available in their repo under the lib/utils/net.py within functions such as compute_diou.

Any suggestions or help on what steps I should take to implement these functions with Detectron2 would be of great help.

Thank you!

**Comment by dhaivat666** [comment]
I have implemented a custom loss function for my purpose. I implemented a loss function in FastRCNNOutputs class. Then when they calculate loss in losses() function within the same class, I call my custom loss function there. Though I am not sure if this the optimal way of doing this or not. But this is what I did and it works decently well. 

So, what you can do is, go in [this file](https://github.com/facebookresearch/detectron2/blob/master/detectron2/modeling/roi_heads/fast_rcnn.py), go to implementation of [FastRCNNOutputs](https://github.com/facebookresearch/detectron2/blob/master/detectron2/modeling/roi_heads/fast_rcnn.py#L126) class, they already have smoothL1loss and crossentropy loss implemented. You can implement your own loss function and call it from losses() function. Hope this helps.

**Comment by ppwwyyxx** [comment]
As suggested by @dhaivat666 , the losses are computed at https://github.com/facebookresearch/detectron2/blob/5e2a1ecccd228227c5a605c0a98d58e1b2db3640/detectron2/modeling/roi_heads/roi_heads.py#L648-L661 so you can implement new ones by either modifying it or write your own `ROIHeads` class (https://detectron2.readthedocs.io/tutorials/write-models.html).

As a note we're looking at possibilities to make the loss computation part of the box head to enable new losses through custom heads - not sure whether/when this will happen though.

**Comment by ferna11i** [comment]
Hi @dhaivat666 and @ppwwyyxx , 
Thank you for pointing out the exact location where I can implement the loss function. I really appreciate it.

There is just one concern I have with the loss function that I am about to implement. The reference code I mentioned in my question uses Detectron where there are two variables bbox_inside_weights and bbox_outside_weights. 

I'm not sure of their functionality yet but I believe the equivalent variables in Detecron2 are: cfg.MODEL.RPN.BBOX_REG_WEIGHTS and cfg.MODEL.ROI_BOX_HEAD.BBOX_REG_WEIGHTS. 

There is a huge possibility that I am wrong and hence would like to get your inputs on where  I should look in the Detectron2 code to find the equivalent variables.


Thank you!

**Comment by ppwwyyxx** [comment]
Those are not equivalent variables, and in fact there are perhaps no equivalent variables of bbox_inside_weights and bbox_outside_weights in detectron2. If you need to implement something new you got to know exactly what you're going to implement - that's not something we can help with.
